File: experiments/run_experiments_NARMAX.m
close all;
clear all;
clc;

addpath(genpath("../offline-baseline"));
addpath(genpath("gen-data"));

%%

% Series of train sizes
trn_sizes = 2.^[6:14];
num_trnsizes = length(trn_sizes);

% Define transient and test indices
transient = 0;
ix_tst = [1:1000] + transient;

options.na = 1; % # output delays
options.nb = 1; % # input delays
options.ne = 1; % # innovation delays
options.nd = 3; % # degree polynomial nonlinearity

M_m = options.na + 1 + options.nb + options.ne;

options.N = 2^16;
options.P = 1;
options.M = 1;

options.fMin = 0;
options.fMax = 100;
options.fs = 1000;
options.type = 'odd';

options.stdu = 1.0;
options.stde = .02;

options.dc = false;
options.crossTerms = true;
options.noiseCrossTerms = true;
options.normalize = false;

% Number of repetitions
num_repeats = 100;

% Preallocate result arrays
results_prd = zeros(num_repeats, num_trnsizes);
results_sim = zeros(num_repeats, num_trnsizes);

r = 1;
while r <= num_repeats
    dbox = waitbar(r/num_repeats);
    
    % Generate signal
    [yTrain, yTest, uTrain, uTest, system] = gen_signal(options);
    
    N_m = size(system.comb,2);
    
    if (max(abs(yTrain)) < 100) && (sum(isnan(yTrain))==0)
    
        % Write signal to file
        save("data/NARMAXsignal_order"+num2str(M_m)+"_N"+num2str(N_m)+"_r" + string(r) + ".mat", "yTrain", "yTest", "uTrain", "uTest", "system", "options")
        
        % Preallocate result arrays
        RMS_prd = zeros(1,num_trnsizes);
        RMS_sim = zeros(1,num_trnsizes);
        
        for n = 1:num_trnsizes
            
            % Establish length of training signal
            ix_trn = [1:trn_sizes(n)] + transient;
        
            % Slice data
            dataTrain.u = uTrain(ix_trn);
            dataTrain.y = yTrain(ix_trn);
            dataTest.u = uTest(ix_tst);
            dataTest.y = yTest(ix_tst);

            % ILS estimator
            [modelNarmaxIter,eNarmaxIter] = fEstPolNarmax(dataTrain,options);

            % 1-step ahead prediction
            [yPredIterTest,ePredIterTest] = fPredPolNarmax(dataTest,modelNarmaxIter);

            % Simulation
            ySimIterTest = fSimPolNarmax(dataTest,modelNarmaxIter);

            % Compute RMS
            RMS_prd(n) = rms(dataTest.y - yPredIterTest);
            RMS_sim(n) = rms(dataTest.y - ySimIterTest);
            
        end
        
        % Write results to file
        save("results/results-NARMAX_ILS_M"+num2str(M_m)+"_N"+num2str(N_m)+"_degree3_r"+num2str(r)+".mat", "RMS_prd", "RMS_sim")
        
        results_prd(r,:) = RMS_prd;
        results_sim(r,:) = RMS_sim;
        
        % Increment repeat
        r = r + 1;
    end    
end
close(dbox) 

% Write results to file
save("results/results-NARMAX_ILS_M"+num2str(M_m)+"_N"+num2str(N_m)+"_degree3.mat", "results_prd", "results_sim")

results_prd(results_prd == Inf) = NaN;
results_sim(results_sim == Inf) = NaN;
results_prd(results_prd > 10) = NaN;
results_sim(results_sim > 10) = NaN;

disp("RMS");
[nanmean(results_prd,1); nanmean(results_sim,1)]

disp("Proportion instable");
[mean(isnan(results_prd)); mean(isnan(results_sim))]

File: experiments/run_experiments_NARMAX.jl
using Revise
using JLD
using MAT
using ProgressMeter
using LinearAlgebra
using ForneyLab
using NARMAX

include("../algorithms/VMP-estimator-NARMAX.jl")
include("../algorithms/RLS-estimator-NARMAX.jl")
include("util.jl")


"""Experimental parameters"""

# System noise std deviation
stde = 0.03

# Series of train sizes
trn_sizes = 2 .^collect(7:10)
num_trnsizes = length(trn_sizes)

# Define transient and test indices
transient = 0
ix_tst = collect(1:1000) .+ transient

# Number of VMP iterations
num_iters = 10

# Number of repetitions
num_repeats = 100

"""Model parameters"""

# Polynomial degree
degree = 3

# Orders
M1 = 1
M2 = 1
M3 = 1
M = 1 + M1 + M2 + M3

# Basis function model
options = Dict()
options["na"] = M1
options["nb"] = M2
options["ne"] = M3
options["nd"] = degree
options["dc"] = false
options["crossTerms"] = true
options["noiseCrossTerms"] = false
PΦ = gen_combs(options)
ϕ(x::Array{Float64,1}) = [prod(x.^PΦ[:,k]) for k = 1:size(PΦ,2)]
N = size(PΦ,2)

# Initialize priors
priors = Dict("θ" => (zeros(N,), Matrix{Float64}(I,N,N)), 
              "τ" => (10, 0.004))
prior_mτ = priors["τ"][1] ./ priors["τ"][2]              

# RLS forgetting factor
λ = 1.00

"""Run experiments"""

# Preallocate results arrays
results_sim_VMP = zeros(num_repeats, num_trnsizes)
results_prd_VMP = zeros(num_repeats, num_trnsizes)
results_sim_RLS = zeros(num_repeats, num_trnsizes)
results_prd_RLS = zeros(num_repeats, num_trnsizes)
results_sim_SYS = zeros(num_repeats, num_trnsizes)
results_prd_SYS = zeros(num_repeats, num_trnsizes)

# Specify model and compile update functions
source_code = model_specification(ϕ, M1=M1, M2=M2, M3=M3, N=N)
eval(Meta.parse(source_code))

println("Starting experiments..")
@showprogress for r = 1:num_repeats

    RMS_sim_VMP = zeros(num_trnsizes,)
    RMS_prd_VMP = zeros(num_trnsizes,)
    RMS_sim_RLS = zeros(num_trnsizes,)
    RMS_prd_RLS = zeros(num_trnsizes,)
    RMS_sim_SYS = zeros(num_trnsizes,)
    RMS_prd_SYS = zeros(num_trnsizes,)
    
    # Read matlab-generated signal
    mat_data = matread("data/NARMAXsignal_stde"*string(stde)*"_pol3_order"*string(M)*"_N"*string(N)*"_r"*string(r)*".mat")

    for n = 1:num_trnsizes

        # Split signals
        ix_trn  = collect(1:trn_sizes[n]) .+ transient
        input_trn = mat_data["uTrain"][ix_trn]
        input_tst = mat_data["uTest"][ix_tst]
        output_trn = mat_data["yTrain"][ix_trn]
        output_tst = mat_data["yTest"][ix_tst]

        # System parameters
        θ_sys = mat_data["system"]["theta"][:]
        τ_sys = inv(mat_data["options"]["stde"]^2)

        # Experiments        
        RMS_sim_VMP[n], RMS_prd_VMP[n] = VMP(input_trn, output_trn, input_tst, output_tst, ϕ, priors, M1=M1, M2=M2, M3=M3, N=N, num_iters=num_iters, computeFE=false)
        RMS_sim_RLS[n], RMS_prd_RLS[n] = RLS(input_trn, output_trn, input_tst, output_tst, ϕ, M1=M1, M2=M2, M3=M3, N=N, λ=λ)
        RMS_sim_SYS[n], RMS_prd_SYS[n] = run_system(input_tst, output_tst, ϕ, θ_sys, M1=M1, M2=M2, M3=M3, N=N)
    end

    # Write results to file
    save("results/results-NARMAX_VMP_stde"*string(stde)*"_pol3_M"*string(M)*"_N"*string(N)*"_degree"*string(degree)*"_mtau_"*string(prior_mτ)*"_r"*string(r)*".jld", "RMS_sim", RMS_sim_VMP, "RMS_prd", RMS_prd_VMP)
    save("results/results-NARMAX_RLS_stde"*string(stde)*"_pol3_M"*string(M)*"_N"*string(N)*"_degree"*string(degree)*"_r"*string(r)*".jld", "RMS_sim", RMS_sim_RLS, "RMS_prd", RMS_prd_RLS)
    save("results/results-NARMAX_SYS_stde"*string(stde)*"_pol3_M"*string(M)*"_N"*string(N)*"_degree"*string(degree)*"_r"*string(r)*".jld", "RMS_sim", RMS_sim_SYS, "RMS_prd", RMS_prd_SYS)

    results_prd_VMP[r,:] = RMS_prd_VMP
    results_sim_VMP[r,:] = RMS_sim_VMP
    results_prd_RLS[r,:] = RMS_prd_RLS
    results_sim_RLS[r,:] = RMS_sim_RLS
    results_prd_SYS[r,:] = RMS_prd_SYS
    results_sim_SYS[r,:] = RMS_sim_SYS

end

# Write results to file
println("Writing results to file..")
save("results/results-NARMAX_VMP_stde"*string(stde)*"_pol3_M"*string(M)*"_N"*string(N)*"_degree"*string(degree)*".jld", "RMS_sim", results_sim_VMP, "RMS_prd", results_prd_VMP)
save("results/results-NARMAX_RLS_stde"*string(stde)*"_pol3_M"*string(M)*"_N"*string(N)*"_degree"*string(degree)*".jld", "RMS_sim", results_sim_RLS, "RMS_prd", results_prd_RLS)
save("results/results-NARMAX_SYS_stde"*string(stde)*"_pol3_M"*string(M)*"_N"*string(N)*"_degree"*string(degree)*".jld", "RMS_sim", results_sim_RLS, "RMS_prd", results_prd_RLS)

println("Experiments complete.")